### PR TITLE
Optimizer: Eliminate joins

### PR DESCRIPTION
**Pull request by barakalon**
Adds optimizer rule to remove unused LEFT JOINs

More context: https://github.com/tobymao/sqlglot/issues/564

### CODE REVIEW

File: sqlglot/optimizer/eliminate_joins.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                unique_outputs.add(select.alias_or_name)
+
+        # All the grouped expressions must be in the output
+        if len(grouped_expressions - grouped_outputs) == 0:
+            return unique_outputs
+
+    return set()
+
+
+def join_condition(join):
```
Comment (barakalon) [inline sqlglot/optimizer/eliminate_joins.py:98]: Extracted from the planner.

Might be a better module to put this...

**Comment by tobymao** [comment]
what about ctes? should we rename the method?

**Comment by barakalon** [comment]
I think that should be a separate rule.
I was going to do it after.

File: sqlglot/optimizer/eliminate_joins.py (hunk 1 above)
Comment (tobymao) [reply sqlglot/optimizer/eliminate_joins.py:96]: how did you find this

File: sqlglot/optimizer/eliminate_joins.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+        unique_outputs = set()
+        for select in scope.selects:
+            output = select.unalias()
+            if output in grouped_expressions:
+                grouped_outputs.add(output)
+                unique_outputs.add(select.alias_or_name)
+
+        # All the grouped expressions must be in the output
+        if len(grouped_expressions - grouped_outputs) == 0:
```
Comment (tobymao) [inline sqlglot/optimizer/eliminate_joins.py:85]: groped_expressions >= grouped_outptus

**Review by tobymao — APPROVED** [review]

File: sqlglot/optimizer/eliminate_joins.py (hunk 1 above)
Comment (barakalon) [reply sqlglot/optimizer/eliminate_joins.py:96]: haha I figured you must've implemented some similar logic with the executor, so I went looking for it. And bam, exactly what I needed.

File: sqlglot/optimizer/eliminate_joins.py (hunk 2 above)
Comment (barakalon) [reply sqlglot/optimizer/eliminate_joins.py:85]: This is a set operation. I'll use explicit set methods so the intent is clearer.

**Pull request merged**